python/hail/dataset.py

```diff
@@ -2326,6 +2326,37 @@ def grm(self):
         jkm = self._jvdf.grm()
         return KinshipMatrix(jkm)
 
+    @handle_py4j
+    @record_method
+    @typecheck(key_expr=str,
+                agg_expr=str,
+                single_key=bool)
+    def group_variants_by(self, key_expr, agg_expr, single_key=True):
+        """
+
+        Group variants by the given key, aggregating entries along samples
+        according to the agg_expr.
+
+        **Examples**
+
+        >>> vds2 = vds.annotateVariantsExpr("va.foo = gs.filter(g => g.isDefined).map(g => g.gt).sum()")
```

Comment:
I don't think the filter does anything here: if g is not defined, then g.gt is not defined and treated as 0 in the sum aggregator. And true is lowercase in next line. How about:

Compute the max depth of each sample on each contig:
 >>> vds = vds.group_variants_by("v.contig", "gs.map(g => g.dp).max()")

I think we should keep the core example of aggregating by gene. You'll need to add a gene (string) and weight (float64) annotation for a small vcf so that a usable vds is created on doc build (i think you can just use the one i made...but it has overlapping genes), and then give the groupBy part of the examples from the 0.1 docs, explaining what will be produced in each case.

'gs.map(g => g.gt).max()'
'gs.map(g => va.weight * g.gt).sum()'
'gs.map(g => va.weight * orElse(g.gt.toDouble, 2 * va.qc.AF)).sum()'

we can discuss